File: 09-Prototype/src/java/chapter09/Rectangle.java
package chapter09;

public class Rectangle extends Shape {
    private int width;
    private int height;

    public Rectangle(int x, int y, String color, int width, int height) {
        super(x, y, color);
        this.width = width;
        this.height = height;
    }

    // Copy constructor
    public Rectangle(Rectangle source) {
        super(source);
        this.width = source.width;
        this.height = source.height;
    }

    @Override
    public Rectangle clone() {
        return new Rectangle(this);
    }

    public int getWidth() { return width; }
    public int getHeight() { return height; }

    @Override
    public String toString() {
        return "Rectangle [" + width + "x" + height + " " + super.toString() + "]";
    }
}

File: 09-Prototype/src/go/rectangle.go
package main

import "fmt"

type Rectangle struct {
	baseShape
	Width  int
	Height int
}

func NewRectangle(x, y int, color string, width, height int) *Rectangle {
	return &Rectangle{
		baseShape: baseShape{X: x, Y: y, Color: color, Tags: []string{}},
		Width:     width,
		Height:    height,
	}
}

func (r *Rectangle) Clone() Shape {
	return &Rectangle{
		baseShape: baseShape{
			X:     r.X,
			Y:     r.Y,
			Color:  r.Color,
			Tags:   deepCopyTags(r.Tags),
		},
		Width:  r.Width,
		Height: r.Height,
	}
}

func (r *Rectangle) String() string {
	return fmt.Sprintf("Rectangle [%dx%d (%d,%d) color=%s tags=%v]",
		r.Width, r.Height, r.X, r.Y, r.Color, r.Tags)
}
